core/storage/btree.rs

```diff
@@ -2360,55 +2361,59 @@ fn free_cell_range(
 /// Defragment a page. This means packing all the cells to the end of the page.
 fn defragment_page(page: &PageContent, usable_space: u16) {
     tracing::debug!("defragment_page");
-    let cloned_page = page.clone();
-    // TODO(pere): usable space should include offset probably
-    let mut cbrk = usable_space;
-
-    // TODO: implement fast algorithm
-
-    let last_cell = usable_space - 4;
-    let first_cell = cloned_page.unallocated_region_start() as u16;
 
-    if cloned_page.cell_count() > 0 {
-        let read_buf = cloned_page.as_ptr();
-        let write_buf = page.as_ptr();
-
-        for i in 0..cloned_page.cell_count() {
-            let (cell_offset, _) = page.cell_pointer_array_offset_and_size();
-            let cell_idx = cell_offset + (i * 2);
-
-            let pc = cloned_page.read_u16_no_offset(cell_idx);
-            if pc > last_cell {
-                unimplemented!("corrupted page");
+    let mut cbrk = usable_space;
+    let first_cell = page.unallocated_region_start() as u16;
+    let n_cell = page.cell_count();
+
+    if n_cell > 0 {
+        let (cell_offset, _) = page.cell_pointer_array_offset_and_size();
+
+        // copy current page data into a temp buffer
+        let mut temp = vec![0u8; usable_space as usize];
+        temp.copy_from_slice(page.buffer.buf_mut().as_mut_slice());
+
+        // collect cell pointers and sizes
+        let mut cells: Vec<(usize, u16, usize)> = (0..n_cell)
+            .map(|i| {
+                let cell_idx = cell_offset + (i * 2);
+                let pc = page.read_u16_no_offset(cell_idx);
+                let (_, size) = page.cell_get_raw_region(
+                    i,
+                    payload_overflow_threshold_max(page.page_type(), usable_space),
+                    payload_overflow_threshold_min(page.page_type(), usable_space),
+                    usable_space as usize,
+                );
+                (i, pc, size)
+            })
+            .collect();
```

Comment:
I don't get this, previously we had a single page copy and now we have page copy and allocation of all cells in the page, essentially this feels like a two copy of the same page? Maybe I'm missing something.